core/cli/src/plugin/options.ts

```diff
@@ -74,48 +74,65 @@ export const substituteOptionTags = (plugin: Plugin, config: ValidPluginsConfig)
-      const error = new ToolKitError('cannot reference plugin options when specifying options')
-      error.details = `YAML tag referencing options used at path '${path.join('.')}'`
-      throw error
+      return invalid([`YAML tag referencing options used at path '${path.join('.')}'`])
     }
+    return valid(undefined)
   }
 
   // recursively walk through the parsed config, searching for the tag
   // identifiers we've inserted during parsing, and substitute them for
   // resolved option values
-  const deeplySubstitute = (node: unknown, path: (string | number)[]): unknown => {
+  const deeplySubstitute = (node: unknown, path: (string | number)[]): Validated<unknown> => {
     if (Array.isArray(node)) {
-      return node.map((item, i) => deeplySubstitute(item, [...path, i]))
+      return reduceValidated(node.map((item, i) => deeplySubstitute(item, [...path, i])))
     } else if (node && typeof node === 'object') {
       const entries = Object.entries(node)
       // !toolkit/option will be marked as a single entry within an object
       // e.g., { foo: { '__toolkit/option__': 'foo.bar' } }
       if (entries[0]?.[0] === toolKitOptionIdent) {
-        validateTagPath(path)
-        const optionPath = entries[0][1] as string
-        return resolveOptionPath(optionPath)
+        return validateTagPath(path).map(() => {
+          const optionPath = entries[0][1] as string
+          return resolveOptionPath(optionPath)
+        })
       } else {
-        const substituted: Record<string, unknown> = {}
+        const substituted: Validated<[string, unknown]>[] = []
         for (const [key, value] of entries) {
           if (key.startsWith(toolKitIfDefinedIdent)) {
-            validateTagPath(path)
+            const validatedKey = validateTagPath(path)
+            if (!validatedKey.valid) {
+              /* eslint-disable-next-line @typescript-eslint/no-explicit-any --
+               * Invalid objects don't need to match the inner type
+               **/
+              substituted.push(validatedKey as Validated<any>)
+            }
             // the option path is concatenated after the !toolkit/if-defined
             // identifier
             const optionPath = key.slice(toolKitIfDefinedIdent.length)
             const optionValue = resolveOptionPath(optionPath)
-            if (optionValue) {
-              Object.assign(substituted, deeplySubstitute(value, path))
-            } // don't include the node if !optionValue
+            // keep walking the path if we've found an error here so we can
+            // gather even more errors to show the user. else skip traversal if
+            // we aren't going to include the node
+            if (optionValue || !validatedKey.valid) {
+              const subbedValues = deeplySubstitute(value, path)
+              if (subbedValues.valid) {
+                substituted.push(...Object.entries(subbedValues.value as object).map((v) => valid(v)))
+              } else {
```

Comment:
**suggestion**: i _think_ this could use `validated.flatMap`?

Reply:
it could use `validated.map` (not `validated.flatMap` because we want `Validated<T>[]` not `Validated<T[]>` here) but I think it makes it less readable in this case